src/computability/tm_computable.lean

```diff
@@ -0,0 +1,220 @@
+/-
+Copyright (c) 2020 Pim Spelier, Daan van Gent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pim Spelier, Daan van Gent.
+-/
+
+import computability.encoding
+import computability.turing_machine
+import data.polynomial.basic
+import data.polynomial.eval
+
+/-!
+# Computable functions
+
+This file contains the definition of a Turing machine with some finiteness conditions
+(bundling the definition of TM2 in turing_machine.lean), a definition of when a TM gives a certain
+output (in a certain time), and the definition of computability (in polytime or any time function)
+of a function between two types that have an encoding (as in encoding.lean).
+
+## Main theorems
+
+- `id_computable_in_poly_time` : a TM + a proof it computes the identity on a type in polytime.
+- `id_computable`              : a TM + a proof it computes the identity on a type.
+
+## Implementation notes
+
+To count the execution time of a Turing machine, we have decided to count the number of times the
+`step` function is used. Each step executes a statement (of type stmt); this is a function, and
+generally contains multiple "fundamental" steps (pushing, popping, so on). However, as functions
+only contain a finite number of executions and each one is executed at most once, this execution
+time is up to multiplication by a constant the amount of fundamental steps.
+-/
+
+open computability
+
+/-- A bundled TM2 (an equivalent of the classical Turing machine, defined starting from
+the namespace `turing.TM2` in `turing_machine.lean`), with an input and output stack,
+ a main function, an initial state and some finiteness guarantees. -/
+structure fin_tm2 :=
+ {K : Type} [K_decidable_eq : decidable_eq K] -- index type of stacks
+ (k₀ k₁ : K) -- input and output stack
+ (Γ : K → Type) -- type of stack elements
+ (Λ : Type) (main : Λ)  -- type of function labels
+ (σ : Type) (initial_state : σ) -- type of states of the machine
+ [σ_fin : fintype σ]
+ [Γk₀_fin : fintype (Γ k₀)]
+ (M : Λ → turing.TM2.stmt Γ Λ σ) -- the program itself, i.e. one function for every function label
+
+namespace fin_tm2
```

Comment:
Maybe this should also be in the `turing` namespace, since it is specifically about turing machines?